compiler/noirc_frontend/src/elaborator/mod.rs

```diff
@@ -1216,19 +1385,46 @@ impl<'context> Elaborator<'context> {
                 self.define_function_metas_for_functions(function_set);
                 self.self_type = None;
                 self.generics.clear();
+                self.generic_idents.clear();
             }
+
+            self.pop_scope();
         }
 
         for trait_impl in trait_impls {
+            self.push_scope();
+
             self.file = trait_impl.file_id;
             self.local_module = trait_impl.module_id;
 
+            trait_impl.trait_id = self.resolve_trait_by_path(trait_impl.trait_path.clone());
             let unresolved_type = &trait_impl.object_type;
+
             self.add_generics(&trait_impl.generics);
             trait_impl.resolved_generics = self.generics.clone();
 
-            let trait_generics =
-                vecmap(&trait_impl.trait_generics, |generic| self.resolve_type(generic.clone()));
+            // Fetch trait constraints here
+            let trait_generics = if let Some(trait_id) = trait_impl.trait_id {
+                let trait_def = self.interner.get_trait(trait_id);
+                let resolved_generics = trait_def.generics.clone();
+                assert_eq!(resolved_generics.len(), trait_impl.trait_generics.len());
+                trait_impl
+                    .trait_generics
+                    .iter()
+                    .enumerate()
+                    .map(|(i, generic)| {
+                        let generic = generic.clone();
+                        if resolved_generics[i].is_numeric_generic {
+                            self.resolve_numeric_type(generic)
+                        } else {
+                            self.resolve_type(generic)
+                        }
```

Comment:
We could move to a design like:
```rs
self.resolve_type_of_kind(generic, resolved_generics[i].kind)
```

Reply:
I moved to your later of `self.resolve_type_inner(generic, resolved_generics[i].kind)`